### PR TITLE
[AppService]: fix #18028- Publish-AzWebapp doesn't handle relative paths properly

### LINKED ISSUE #18028: WebApp:Publish-AzWebapp doesn't handle relative paths properly
### Description

Public-AzWebapp interprets relative paths incorrectly.

Any relative path I use is appended to c:\windows\system32.  There was a previous report of this which was closed without a proper explanation.

PS > Publish-AzWebApp -ResourceGroupName POC -Name Test -ArchivePath ".\TestApp\TestApp.zip" -Force
Publish-AzWebApp: Could not find a part of the path 'C:\Windows\system32\TestApp\TestApp.zip'.

### Issue script & Debug output

```PowerShell
PS c:\Users\XXX\src> Publish-AzWebApp -ResourceGroupName POC -Name Test -ArchivePath ".\TestApp\TestApp.zip" -Force
Publish-AzWebApp: Could not find a part of the path 'C:\Windows\system32\TestApp\TestApp.zip'.
```

### Environment data

```PowerShell
Name                           Value
----                           -----
PSVersion                      7.2.3
PSEdition                      Core
GitCommitId                    7.2.3
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Module versions

```PowerShell
ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Script     2.2.0                 Az.Websites                         {Add-AzWebAppAccessRestrictionRule, Add-AzWebAppTrafficRouting, Edit-AzWebAppBackupConfiguration, Enter-AzWebAppContainerPSSession…}
```

### Error output

```PowerShell
HistoryId: 206

Message        : Could not find a part of the path 'C:\Windows\system32\TestApp\TestApp.zip'.
StackTrace     :    at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
                    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
                    at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
                    at System.IO.File.OpenRead(String path)
                    at Microsoft.Azure.Commands.WebApps.Cmdlets.WebApps.PublishAzureWebAppCmdlet.<>c__DisplayClass13_0.<ExecuteCmdlet>b__0()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ConfirmAction(Boolean force, String continueMessage, String processMessage, String target, Action action, Func`1 useShouldContinue)
                    at Microsoft.Azure.Commands.ResourceManager.Common.AzureRMCmdlet.ConfirmAction(Boolean force, String continueMessage, String processMessage, String target, Action action)
                    at Microsoft.Azure.Commands.WebApps.Cmdlets.WebApps.PublishAzureWebAppCmdlet.ExecuteCmdlet()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.<>c__3`1.<ExecuteSynchronouslyOrAsJob>b__3_0(T c)
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.ExecuteSynchronouslyOrAsJob[T](T cmdlet, Action`1 executor)
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.CmdletExtensions.ExecuteSynchronouslyOrAsJob[T](T cmdlet)
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
Exception      : System.IO.DirectoryNotFoundException
InvocationInfo : {Publish-AzWebApp}
Line           : Publish-AzWebApp -ResourceGroupName POC -Name Test -ArchivePath ".\TestApp\TestApp.zip" -Force
Position       : At line:1 char:1
                 + Publish-AzWebApp -ResourceGroupName POC -Name Test -Arc …
                 + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
HistoryId      : 206
```

### PR DESCRIPTION
**Pull request by Kotasudhakarreddy**
## Description
#18028

Public-AzWebapp interprets relative paths incorrectly.

Any relative path I use is appended to c:\windows\system32. There was a previous report of this which was closed without a proper explanation.

PS > Publish-AzWebApp -ResourceGroupName POC -Name Test -ArchivePath ".\TestApp\TestApp.zip" -Force
Publish-AzWebApp: Could not find a part of the path 'C:\Windows\system32\TestApp\TestApp.zip'.
## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from the scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** change version of module in pull request

### CODE REVIEW

File: src/Websites/Websites/ChangeLog.md [hunk 1]
```diff
@@ -18,7 +18,7 @@
         - Additional information about change #1
 -->
 ## Upcoming Release
-
+* Fixed `Publish-AzWebapp` to handle relative paths properly [#18028]
```
Comment (BethanyZhou) [inline src/Websites/Websites/ChangeLog.md:21]:
```suggestion
* Fixed `Publish-AzWebapp` to handle relative paths properly [#18028]

```

**Review by BethanyZhou — DISMISSED** [review]
.

File: src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs [hunk 2]
```diff
@@ -74,6 +74,8 @@ public override void ExecuteCmdlet()
 
             Action zipDeployAction = () =>
             {
+                if (!Path.IsPathRooted(ArchivePath))
+                    ArchivePath = (this.SessionState.Path.CurrentFileSystemLocation).Path + ArchivePath.Split(new string[] { ".." }, StringSplitOptions.None)[1];
```
Comment (BethanyZhou) [inline src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: There is a `AsJob` parameter in `Publish-AzWebApp`. If `AsJob` is present, method ExecuteCmdlet() will run as a separate thread. Not sure your code whether works in this scenario. Could you give it a test before I merge your code?

Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: Yes, it is working well with Job as well. PFB screenshot FYR. And when I use the `TryResolvePath` method you have suggested, it is failing to execute as Job.

image

Comment (BethanyZhou) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: what is the absolute path of 12509issues.zip

Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs]: yes, you are correct. The absolute path here is `D:\SampleApps\12509issue.zip` .

Comment (BethanyZhou) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs]: Could you try my recommended way and use different locations as test cases?

Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs]: it works only with the `CurrentFileSystemLocation` i.e current directory of the `command prompt ` followed by the given relative path. example:
`D:\>> -ArchivePath '.\SampleApp.Zip\` is equivalent to `D:\SampleApp.Zip`
`D:\>> -ArchivePath '.\Test\SampleApp.Zip\` is equivalent to `D:\Test\SampleApp.Zip`
`D:\>> -ArchivePath '..\Test\SampleApp.Zip\` is equivalent to `D:\Test\SampleApp.Zip`

I have tested all the above scenarios and it is working perfectly.

Comment (BethanyZhou) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs]: How about going to D:\test\> -ArchivePath '.\SampleApp.Zip?

File: src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs [hunk 3]
```diff
@@ -74,6 +74,8 @@ public override void ExecuteCmdlet()
 
             Action zipDeployAction = () =>
             {
+                if (!Path.IsPathRooted(ArchivePath))
+                    ArchivePath = (this.SessionState.Path.CurrentFileSystemLocation).Path + ArchivePath;
```
Comment (BethanyZhou) [inline src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]:
```suggestion
                    ArchivePath = Path.Join(this.SessionState.Path.CurrentFileSystemLocation).Path, ArchivePath);
```

Comment (BethanyZhou) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: I believe this will hand ../Sample/*.Zip as well.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs (hunk 3 above)
Comment (Kotasudhakarreddy) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: `Path.Join()` is not available in .NetStandard.

Comment (BethanyZhou) [reply src/Websites/Websites/Cmdlets/WebApps/PublishAzureWebApp.cs:78]: Alternatively, we can use Path.Combine()

**Review by BethanyZhou — APPROVED** [review]

**Pull request merged**